File: leetcode/python/0067. Add Binary.py
# Time complexity: O(N)
# Space complexity: O(N)


class Solution(object):
    def addBinary(self, a, b):
        """
        :type a: str
        :type b: str
        :rtype: str
        """
        s = self.binary2dec(a) + self.binary2dec(b)
        return self.dectobinary(s)

    def binary2dec(self, bin):
        currExp = 1
        res = 0
        for i in range(len(bin)-1, -1, -1):
            if bin[i] == "1":
                res += currExp
            currExp *= 2
        return res

    def dectobinary(self, dec):
        res = []
        while dec > 0:
            if dec % 2 == 0:
                res.append("1")
            else:
                res.append("0")
            dec /= 2
        return "".join(res)

File: leetcode/java/0067. Add Binary.java
package leetcode.java;

// Time complexity: O(N)
// Space complexity: O(N)

class Solution {
    public String addBinary(String a, String b) {
        StringBuilder sb = new StringBuilder();

        int indexA = a.length() - 1;
        int indexB = b.length() - 1;

        int carry = 0;

        while (indexA >= 0 || indexB >= 0) {
            int currSum = carry;

            if (indexA >= 0) {
                currSum += (a.charAt(indexA) - '0');
                indexA -= 1;
            }

            if (indexB >= 0) {
                currSum += (b.charAt(indexB) - '0');
                indexB -= 1;
            }

            sb.insert(0, currSum % 2);
            carry = currSum / 2;
        }

        if (carry > 0) {
            sb.insert(0, 1);
        }
        return sb.toString();
    }
}